src/vgr_cairo.ml

```diff
@@ -0,0 +1,437 @@
+  | Rot a -> Cairo.rotate s.ctx a; M3.rot2 a
+  | Scale sv -> Cairo.scale s.ctx (V2.x sv) (V2.y sv); M3.scale2 sv
+  | Matrix m -> Cairo.transform s.ctx (cairo_matrix_of_m3 m); m
+  in
+  s.gstate.g_tr <- M3.mul s.gstate.g_tr m
+
+let set_outline s o =
+  if s.gstate.g_outline == o then () else
+  let old = s.gstate.g_outline in
+  s.gstate.g_outline <- o;
+  if old.P.width <> o.P.width then
+    (Cairo.set_line_width s.ctx o.P.width);
+  if old.P.cap <> o.P.cap then
+    (Cairo.set_line_cap s.ctx (cairo_cap o.P.cap));
+  if old.P.join <> o.P.join then
+    (Cairo.set_line_join s.ctx (cairo_join o.P.join));
+  if old.P.miter_angle <> o.P.miter_angle then
+    (Cairo.set_miter_limit s.ctx (Vgr.Private.P.miter_limit o));
+  if old.P.dashes <> o.P.dashes then set_dashes s o.P.dashes;
+  ()
+
+let get_primitive s p = try Hashtbl.find s.prims p with
+| Not_found ->
+    let add_stop g (t, c) =
+      let c = Color.to_srgb c in
+      Cairo.Pattern.add_color_stop_rgba g ~ofs:t
+        (V4.x c) (V4.y c) (V4.z c) (V4.w c) in
+    let create = function
+    | Const c ->
+        let c = Color.to_srgb c in
+        Pattern V4.(Cairo.Pattern.create_rgba (x c) (y c) (z c) (w c))
+    | Axial (stops, pt, pt') ->
+        let g = V2.(Cairo.Pattern.create_linear (x pt)  (y pt)
+                                                (x pt') (y pt')) in
+        List.iter (add_stop g) stops; Pattern g
+    | Radial (stops, f, c, r) ->
+        let g = V2.(Cairo.Pattern.create_radial
+                      (x f) (y f) 0.0 (x c) (y c) r) in
+        List.iter (add_stop g) stops; Pattern g
+    | Raster _ -> assert false
+    in
+    let prim = create p in
+    Hashtbl.add s.prims p prim; prim
+
+let get_font s font = try Hashtbl.find s.fonts font with
+| Not_found ->
+    let cairo_font =
+      let slant = match font.Font.slant with
+      | `Italic -> Cairo.Italic
+      | `Normal -> Cairo.Upright
+      | `Oblique -> Cairo.Oblique in
+      let weight = match font.Font.weight with
+      | `W700 | `W800 | `W900 -> Cairo.Bold
+      | _ -> Cairo.Normal in
+      Font (Cairo.Font_face.create ~family:font.Font.name slant weight)
+    in
+    Hashtbl.add s.fonts font cairo_font; cairo_font
+
+let set_source s p =
+  let (Pattern g) as p = get_primitive s p in
+  Cairo.set_source s.ctx g;
+  p
+
+let set_stroke s p = s.gstate.g_stroke <- set_source s p
+
+let set_fill s p = s.gstate.g_fill <- set_source s p
+
+let set_font s font size =
+  let Font f = get_font s font in
+  Cairo.Font_face.set s.ctx f;
+  Cairo.set_font_size s.ctx size
+
+let set_path s p =
+  let rec loop last = function
+  | [] -> ()
+  | seg :: segs ->
+      match seg with
+      | `Sub pt -> P2.(Cairo.move_to s.ctx (x pt) (y pt)); loop pt segs
+      | `Line pt -> P2.(Cairo.line_to s.ctx (x pt) (y pt)); loop pt segs
+      | `Qcurve (q, pt) ->
+          let x,y = Cairo.Path.get_current_point s.ctx in
+          let p0 = V2.v x y in
+          let c  = V2.((q + 2. * p0) / 3.) in
```

Comment:
Fixed by https://github.com/dbuenzli/vg/commit/fa42d177d2c6eee340816bd5b2230a4f9af38d2e. We get the following animated gif some small discrepancies are still visible, but that is somehow expected.

![diff2](https://cloud.githubusercontent.com/assets/485596/9220754/9fbada26-40dc-11e5-9cb3-af6f99a9f505.gif)